src/Interpreters/Access/InterpreterCheckGrantQuery.cpp

```diff
@@ -0,0 +1,114 @@
+#include <Access/Role.h>
+#include <Access/RolesOrUsersSet.h>
+#include <Access/User.h>
+#include <Interpreters/Context.h>
+#include <Interpreters/removeOnClusterClauseIfNeeded.h>
+#include <Interpreters/QueryLog.h>
+#include <Interpreters/executeDDLQueryOnCluster.h>
+#include <boost/range/algorithm/copy.hpp>
+#include <boost/range/algorithm/set_algorithm.hpp>
+#include <boost/range/algorithm_ext/erase.hpp>
+#include <Interpreters/DatabaseCatalog.h>
+#include "Databases/IDatabase.h"
+#include "Storages/IStorage.h"
+
+namespace DB
+{
+
+namespace
+{
+    /// Extracts access rights elements which are going to check grant
+    void collectAccessRightsElementsToGrantOrRevoke(
+        const ASTCheckGrantQuery & query,
+        AccessRightsElements & elements_to_check_grant)
+    {
+        elements_to_check_grant.clear();
+        /// GRANT
+        elements_to_check_grant = query.access_rights_elements;
+    }
+}
+
+
+BlockIO InterpreterCheckGrantQuery::execute()
+{
+    auto & query = query_ptr->as<ASTCheckGrantQuery &>();
+    query.access_rights_elements.eraseNonGrantable();
+
+    auto current_user_access = getContext()->getAccess();
+
+    /// Collect access rights elements which will be checked.
+    AccessRightsElements elements_to_check_grant;
+    collectAccessRightsElementsToGrantOrRevoke(query, elements_to_check_grant);
+
+
+    /// Replacing empty database with the default. This step must be done before replication to avoid privilege escalation.
+    String current_database = getContext()->getCurrentDatabase();
+    elements_to_check_grant.replaceEmptyDatabase(current_database);
+    query.access_rights_elements.replaceEmptyDatabase(current_database);
+    auto *logger = &::Poco::Logger::get("CheckGrantQuery");
+
+    /// Check If Table/Columns exist.
+    for (const auto & elem : elements_to_check_grant)
+    {
+        try
+        {
+            DatabasePtr database;
+            database = DatabaseCatalog::instance().getDatabase(elem.database);
+            if (!database->isTableExist(elem.table, getContext()))
+            {
+                /// Table not found.
+                return executeQuery("SELECT 0 AS CHECK_GRANT", getContext(), QueryFlags{.internal = true}).second;
+            }
+            auto table = database->getTable(elem.table, getContext());
+
+            auto column_name_with_sizes = table->getColumnSizes();
+            for (const auto & elem_col : elem.columns)
+            {
+                bool founded = false;
+                for (const auto & col_in_table : column_name_with_sizes)
+                {
+                    if (col_in_table.first == elem_col)
+                    {
+                        founded = true;
+                        break;
+                    }
+                }
+                if (!founded)
+                {
+                    /// Column not found.
+                    return executeQuery("SELECT 0 AS CHECK_GRANT", getContext(), QueryFlags{.internal = true}).second;
```

Comment:
Calling `executeQuery` is a quite heavy way to just return `0` as a result.
It's much better to just build the result block here from scratch - in the same way how it's built in `InterpreterExistsQuery`:

```
    return QueryPipeline(std::make_shared<SourceFromSingleChunk>(Block{{
        ColumnUInt8::create(1, result),
        std::make_shared<DataTypeUInt8>(),
        "result" }}));
```

Reply:
> Calling `executeQuery` is a quite heavy way to just return `0` as a result.
> 
> It's much better to just build the result block here from scratch - in the same way how it's built in `InterpreterExistsQuery`:
> 
> ```
>     return QueryPipeline(std::make_shared<SourceFromSingleChunk>(Block{{
>         ColumnUInt8::create(1, result),
>         std::make_shared<DataTypeUInt8>(),
>         "result" }}));
> ```

Got it.